### PR TITLE
[Op]: Added SearchSorted op def.

### PR DESCRIPTION
**Pull request by pkowalc1**
### Details:
 - Added SearchSorted op def with unittests.

### Tickets:
 - *CVS-154141*

NOTE: Depends on https://github.com/openvinotoolkit/openvino/pull/26887

### CODE REVIEW

File: src/core/shape_inference/include/search_sorted_shape_inference.hpp [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+namespace ov {
+namespace op {
+namespace v15 {
+template <class TShape, class TRShape = result_shape_t<TShape>>
+std::vector<TRShape> shape_infer(const SearchSorted* op, const std::vector<TShape>& input_shapes) {
+    const auto& sorted_shape = input_shapes[0];
+    const auto& values_shape = input_shapes[1];
+    auto output_shape = values_shape;
+    TShape::merge_into(output_shape, sorted_shape);
```
Comment (praasz) [inline src/core/shape_inference/include/search_sorted_shape_inference.hpp:18]: Assert the result e.g.:
```suggestion
    NODE_SHAPE_INFER_CHECK(op, TShape::merge_into(output_shape, sorted_shape), input_shapes, "Shapes are not compatible ....");
```
Is required when static shapes used (CPU).

And `merge_into` do same validation as [here](https://github.com/openvinotoolkit/openvino/pull/26904/files#diff-591173189ba331aa8f2ee9d6e6f3fa6fcae26dd755a5adeec3311ef107b7353aR34)

File: src/core/include/openvino/op/search_sorted.hpp [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    bool get_right_mode() const {
+        return m_right_mode;
+    }
+
+    void set_right_mode(bool right_mode) {
+        m_right_mode = right_mode;
+    }
+
+private:
+    void validate();
```
Comment (praasz) [inline src/core/include/openvino/op/search_sorted.hpp:40]:
```suggestion
    void validate() const;
```
?

File: src/core/tests/type_prop/search_sorted.cpp
Comment (praasz) [inline src/core/tests/type_prop/search_sorted.cpp:1]: Add test(s) in `src/core/tests/visitors/op/`

File: src/core/tests/type_prop/search_sorted.cpp [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+                             const PartialShape& expected_output_shape) {
+    auto sorted = make_shared<op::v0::Parameter>(element::i32, sorted_shape);
+    auto values = make_shared<op::v0::Parameter>(element::i32, values_shape);
+    auto search_sorted_op = make_shared<op::v15::SearchSorted>(sorted, values);
+    EXPECT_EQ(search_sorted_op->get_element_type(), element::i64);
+    EXPECT_EQ(search_sorted_op->get_output_partial_shape(0), expected_output_shape);
+}
+
+TEST(type_prop, search_sorted_shape_infer_equal_inputs) {
+    PerformShapeTest({1, 3, 6}, {1, 3, 6}, {1, 3, 6});
```
Comment (praasz) [inline src/core/tests/type_prop/search_sorted.cpp:33]: Add test with scalar shapes ?

File: src/core/include/openvino/op/search_sorted.hpp [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+    void validate_and_infer_types() override;
+    bool visit_attributes(AttributeVisitor& visitor) override;
+    std::shared_ptr<Node> clone_with_new_inputs(const OutputVector& new_args) const override;
+
+    bool get_right_mode() const {
+        return m_right_mode;
+    }
+
+    void set_right_mode(bool right_mode) {
```
[the remark this answers is not in the document]
Comment (pkowalc1) [reply src/core/include/openvino/op/search_sorted.hpp:35]: Following the convention of other operators - all of them have setters and getters.

File: src/core/src/op/search_sorted.cpp [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+                              sorted_shape.rank().get_length() == values_shape.rank().get_length(),
+                              "Sorted sequence and values have different ranks.");
+
+        for (int64_t i = 0; i < sorted_shape.rank().get_length() - 1; ++i) {
+            NODE_VALIDATION_CHECK(this,
+                                  sorted_shape[i].compatible(values_shape[i]),
+                                  "Sorted sequence and values has different ",
+                                  i,
+                                  " dimension.");
+        }
```
[the remark this answers is not in the document]
Comment (pkowalc1) [reply src/core/src/op/search_sorted.cpp:45]: This is a very bad approach - mixing validation and shape infer inside shape inference function - so I don't want to follow it unless absolutely necessary. Even then I will rather mark it as a hack.

File: src/core/tests/type_prop/search_sorted.cpp (hunk 3 above)
Comment (pkowalc1) [reply src/core/tests/type_prop/search_sorted.cpp:33]: By 'scalar shape', you meant 1d shape or shape describing just one value?(eg. {1}. {1,1,1,1}, etc)?

File: src/core/tests/type_prop/search_sorted.cpp [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    try {                                                             \
+        auto op = make_shared<op::v15::SearchSorted>(SORTED, VALUES); \
+        FAIL() << "Exception not thrown";                             \
+    } catch (const NodeValidationFailure& error) {                    \
+        EXPECT_THAT(error.what(), testing::HasSubstr(SUBSTRING));     \
+    } catch (...) {                                                   \
+        FAIL() << "Unexpected exception thrown";                      \
+    }
+
+static void PerformShapeTest(const PartialShape& sorted_shape,
```
[the remark this answers is not in the document]
Comment (pkowalc1) [reply src/core/tests/type_prop/search_sorted.cpp:22]: Considered, but will leave it as it is.

File: src/core/include/openvino/op/search_sorted.hpp (hunk 2 above)
Comment (pkowalc1) [reply src/core/include/openvino/op/search_sorted.hpp]: Yes, should be normally const, but other methods, where this function is used(validate_and_infer_types()) are not const.

File: src/core/tests/type_prop/search_sorted.cpp [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+TEST(type_prop, search_sorted_shape_infer_both_dynamic_4) {
+    PerformShapeTest({-1, -1, 50}, {-1, -1, 20}, {-1, -1, 20});
+}
+
+TEST(type_prop, search_sorted_shape_infer_different_types) {
+    auto sorted = make_shared<ov::op::v0::Parameter>(element::f32, Shape{1, 3, 6});
+    auto values = make_shared<ov::op::v0::Parameter>(element::i32, Shape{1, 3, 6});
+    EXPECT_THROW_SUBSTRING(values, sorted, std::string("must have the same element type"));
```
[the remark this answers is not in the document]
Comment (pkowalc1) [reply src/core/tests/type_prop/search_sorted.cpp:67]: Fixed.

File: src/core/tests/type_prop/search_sorted.cpp (hunk 3 above)
Comment (praasz) [reply src/core/tests/type_prop/search_sorted.cpp:33]: Scalar shape is `Shape{}` has no dimensions. Maybe it should be not allowed for inference.

File: src/core/tests/type_prop/search_sorted.cpp
Comment (pkowalc1) [reply src/core/tests/type_prop/search_sorted.cpp:1]: Fixed.

File: src/core/shape_inference/include/search_sorted_shape_inference.hpp [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+    if (output_shape.rank().is_static()) {
+        auto last_it = output_shape.end() - 1;
+        if (values_shape.rank().is_static()) {
+            *last_it = *(input_shapes[1].end() - 1);
+        } else {
+            *last_it = Dimension::dynamic();
+        }
```
[the remark this answers is not in the document]
Comment (pkowalc1) [reply src/core/shape_inference/include/search_sorted_shape_inference.hpp:28]: If you see a better(and fast solution)-please propose it. Refactoring merge_into is out of the scope of this pr.

File: src/core/src/op/search_sorted.cpp [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+SearchSorted::SearchSorted(const Output<Node>& sorted_sequence, const Output<Node>& values, bool right_mode)
+    : Op({sorted_sequence, values}),
+      m_right_mode(right_mode) {
+    constructor_validate_and_infer_types();
+}
+
+void SearchSorted::infer_type() {
+    const auto& output_shapes = shape_infer(this, ov::util::get_node_input_partial_shapes(*this));
+    set_output_type(0, ov::element::i64, output_shapes[0]);
```
[the remark this answers is not in the document]
Comment (pkowalc1) [reply src/core/src/op/search_sorted.cpp]: Honestly I would like to add it only if really needed, pytorch only supports i32 via special flag which is not used in reference model that this op is needed for.

File: src/core/src/op/search_sorted.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (pkowalc1) [reply src/core/src/op/search_sorted.cpp]: Edit: validate moved to shape_infer.

File: src/core/tests/type_prop/search_sorted.cpp (hunk 3 above)
Comment (pkowalc1) [reply src/core/tests/type_prop/search_sorted.cpp:29]: If you ask for fully dynamic case, where even rank is not known, then there is already a test case for it.

**Comment by pkowalc1** [comment]
Dear reviewers, I tried to apply all of your comments, which I have considered valid. If you see any other serious issues(blockers), please point it out. Otherwise please pass this review since the deadline for adding the op is 18th Oct and it is already at risk.

File: src/core/src/op/search_sorted.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (mitruska) [reply src/core/src/op/search_sorted.cpp]: The updated solution ("op->validate()" call inside the shape_infer) doesn't validate the shapes provided as `input_shapes` argument (ie.g. static shapes of Tensors), but the original op input shapes which can be still dynamic.
https://github.com/openvinotoolkit/openvino/blob/cd3873d0f966be1deafab1bff4d939270fdfdf21/src/core/shape_inference/include/search_sorted_shape_inference.hpp#L14-L16

**Review by mitruska — APPROVED** [review]
Approved, with agreement to apply `shape_infer` improvements in a separate PR.
 (https://github.com/openvinotoolkit/openvino/pull/26904#discussion_r1791676700 - to be resolved within the ticket 154368).

File: src/core/src/op/search_sorted.cpp [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+bool SearchSorted::validate() const {
+    NODE_VALIDATION_CHECK(this, get_input_size() == 2);
+    NODE_VALIDATION_CHECK(this,
+                          get_input_element_type(0) == get_input_element_type(1),
+                          "Sorted sequence and values must have the same element type.");
+
+    const auto& sorted_shape = get_input_partial_shape(0);
+    const auto& values_shape = get_input_partial_shape(1);
+
+    if (sorted_shape.rank().is_static() && values_shape.rank().is_static() && sorted_shape.rank().get_length() > 1) {
```
Comment (praasz) [inline src/core/src/op/search_sorted.cpp:30]:
```suggestion
    if (sorted_shape.rank().is_static() && values_shape.rank().is_static() && sorted_shape.rank().get_length() > 0) {
```

File: src/core/tests/type_prop/search_sorted.cpp (hunk 3 above)
Comment (praasz) [reply src/core/tests/type_prop/search_sorted.cpp:29]: I mean static case when shape has rank 0. Only one element but without dimensions.

**Review by praasz — APPROVED** [review]

File: src/core/src/op/search_sorted.cpp (hunk 10 above)
Comment (pkowalc1) [reply src/core/src/op/search_sorted.cpp:30]: No, 1d case is a special case, that's why that check is added.

**Pull request merged**